lib/minfraud/transaction.rb

```diff
@@ -28,60 +32,139 @@ class Transaction
-      attrs = [:ip, :city, :state, :postal, :country, :license_key, :ship_addr, :ship_city, :ship_state, :ship_postal, :ship_country, :email_domain, :email_md5, :phone, :bin, :session_id, :user_agent, :accept_language, :txn_id, :amount, :currency, :txn_type, :avs_result, :cvv_result, :requested_type, :forwarded_ip, :shop_id]
-      attrs.map! do |a|
-        [a, send(a)]
-      end
-      Hash[attrs]
+      Hash[ATTRIBUTES.map { |a| [a, send(a)] }].compact
     end
 
     private
 
     # Ensures the required attributes are present
     # @return [Boolean]
     def has_required_attributes?
-      ip && txn_id && license_key
+      REQUIRED_ATTRIBUTES.none? { |attr| send(attr).nil? || send(attr).empty? }
     end
 
     # Validates the types of the attributes
-    # @return [nil, TransactionError]
+    # @raise [TransactionError] if present attributes are not valid
+    # @return [nil]
     def validate_attributes
-      [:ip, :city, :state, :postal, :country, :txn_id, :license_key].each { |s| validate_string(s) }
+      INPUT_STRING_ATTRIBUTES.each { |attr| validate_string(attr) }
+      INPUT_NUMERIC_ATTRIBUTES.each { |attr| validate_number(attr) }
+      CONVERT_STRING_ATTRIBUTES.each { |attr| convert_to_string(attr) }
+      validate_cvv
     end
 
     # Given the symbol of an attribute that should be a string,
     # it checks the attribute's type and throws an error if it's not a string.
     # @param attr_name [Symbol] name of the attribute to validate
-    # @return [nil, TransactionError]
+    # @raise [TransactionError] if attribute is not a string
+    # @return [nil]
     def validate_string(attr_name)
-      attribute = self.send(attr_name)
+      attribute = send(attr_name)
       if attribute && !attribute.instance_of?(String)
         raise TransactionError, "Transaction.#{attr_name} must be a string"
       end
     end
 
+    # Given the symbol of an attribute that should be a number,
+    # it checks the attribute's type and throws an error if it
+    # cannot be interpreted as a BigDecimal
+    # @param attr_name [Symbol] name of the attribute to validate
+    # @raise [TransactionError] if attribute is not a number
+    # @return [nil]
+    def validate_number(attr_name)
+      attribute = send(attr_name)
+      return if attribute.nil? || attribute.is_a?(Numeric)
+      begin
+        BigDecimal(attribute)
+      rescue
+        raise TransactionError, "Transaction.#{attr_name} must be a number"
+      end
+    end
+
+    # Validates the cvv_result, which must be a single character
+    # @raise [TransactionError] if cvv_result is not a single character
+    # @return [nil]
+    def validate_cvv
+      return nil if cvv_result.nil? || cvv_result.empty?
+      unless cvv_result.length == 1
+        raise TransactionError, "Transaction.cvv_result must be a single letter"
+      end
+    end
+
+    # Converts the given attribute to a string
+    # @param attr_name [Symbol] name of the attribute to convert
+    # @return [nil]
```

Comment:
`@return [void]` might be more representative of what the method does since all it does is assigning an attribute 🙂